runtime/near-vm-runner/RUNTIMES.md

```diff
@@ -0,0 +1,176 @@
+specification also permits to have multiple different bit representations and goes out of its way
+to specify the sign of the `NaN` result to be non-deterministic. The determinism and
+reproducibility requirements of a NEAR validator make it necessary for the runtime to ensure NaN
+signs and payloads are always deterministic.
+
+Another detail to keep an eye out is whether the floating point results are being rounded
+correctly. It is well known that the x87 co-processor does not necessarily round the results
+correctly, so a special attention to the implementation of the specification on the 32-bit x86
+targets is warranted here.
+
+## Security
+
+NEAR protocol validators execute arbitrary Wasm code submitted by the participants of the network,
+much like e.g. web browsers execute untrusted JavaScript code. It is critical that the runtime
+executes the Wasm code in its own isolated environment. The contract code must not be presented
+with an opportunity to access resources from the system running a validator node outside of what is
+enabled via host functions exposed by the validator implementation itself.
+
+Additionally, contracts should be isolated from each other as well. Never should there be a
+situation where it is possible for a contract to recover and extradite state of a different
+contract.
+
+A bug in the runtime is most likely to take a form of some sort of memory management issue, where
+the memory is reused by the contract without zeroing it, or where the memory accesses aren't
+validated correctly and allow access outside of the memory regions allocated for the contract.
+
+## Reliability
+
+The required and desirable properties of a runtime implementation must remain true while the
+changes are made to it. Compilers and, by extension, runtimes are complex systems and a poor
+architecture may result in even simple changes introducing correctness or security bugs, which
+aren't necessarily easily discovered.
+
+A reliable code base in this sense should be architected in such a way that the algorithms they
+implement and the way they execute Wasm code remains correct by construction and is easy to verify.
+
+Size of a test suite, architecture of the code base and the number of developers/users are all
+reasonably correlated signals as to the reliability which one can expect from a runtime.
+
+## Platform support
+
+The NEAR validator implementation currently targets `x86_64-linux`, so stellar support for this
+target is a hard requirement.
+
+However, developers may not necessarily have a `x86_64` machine to develop with, or they might be
+using a device running a different operating system. With the recent surge in popularity of devices
+based on the Arm CPU architecture chips and generally high popularity of the macOS and Windows
+operating systems it would be a huge boon to the development experience if the runtime used also
+supported these other targets.
+
+## Runtime performance
```

Comment:
Let's maybe split this into two primary metrics we care about: 

* *latency* to run a simple function in a big contract (rationale: many contracts are relatively simple in terms of business logic and do little compute)
* *throughput* to run a relatively long wasm computation as fast as possible (rationale: at the same time, there are some compute-heavy contracts, especially those that run interpreters in wasm)

Reply:
In general, I feel like our runtime model has three parts to it: 

* compilation (read wasm, compile to machine code, write machine code to disk)
* linking/loading (load machine code from disk, load it into the executable memory, unload it after execution)
* actual execution

Reply:
I attempted rewording the sections around performance to introduce more detailed view of the requirements and the parts that go into it.